snuba/pipeline/composite_entity_processing.py

```diff
@@ -0,0 +1,178 @@
+from snuba.pipeline.composite_storage_processing import check_sub_query_storage_sets
+from snuba.query import ProcessableQuery
+from snuba.query.composite import CompositeQuery
+from snuba.query.data_source.join import IndividualNode, JoinClause, JoinVisitor
+from snuba.query.data_source.simple import Entity, Table
+from snuba.query.data_source.visitor import DataSourceVisitor
+from snuba.query.joins.equivalence_adder import add_equivalent_conditions
+from snuba.query.joins.subquery_generator import generate_subqueries
+from snuba.query.logical import Query as LogicalQuery
+from snuba.query.query_settings import QuerySettings
+
+
+def translate_composite_query(
+    query: CompositeQuery[Entity], query_settings: QuerySettings
+) -> CompositeQuery[Table]:
+    """
+    A function that traverses a logical composite query, applies all entity processors,
+    and translates the result into a physical composite query.
+    """
+    add_equivalent_conditions(query)
+    generate_subqueries(query)
+    physical_query = _translate_logical_composite_query(query, query_settings)
+    return physical_query
+
+
+def _translate_logical_composite_query(
+    query: CompositeQuery[Entity], settings: QuerySettings
+) -> CompositeQuery[Table]:
+    """
+    Given a logical composite query, this function traverses each sub-query node,
+    executes all entity processors associated with each sub-query node,
+    and builds a physical composite query.
+    """
+
+    translated_source = CompositeDataSourceTransformer(settings).visit(
+        query.get_from_clause()
+    )
+
+    return CompositeQuery(
+        from_clause=translated_source,
+        selected_columns=query.get_selected_columns(),
+        array_join=query.get_arrayjoin(),
+        condition=query.get_condition(),
+        groupby=query.get_groupby(),
+        having=query.get_having(),
+        order_by=query.get_orderby(),
+        limitby=query.get_limitby(),
+        limit=query.get_limit(),
+        offset=query.get_offset(),
+        totals=query.has_totals(),
+        granularity=query.get_granularity(),
+    )
+
+
+class CompositeDataSourceTransformer(
+    DataSourceVisitor[
+        Union[
+            ProcessableQuery[Table],
+            CompositeQuery[Table],
+            JoinClause[Table],
+        ],
+        Entity,
+    ]
+):
+    """
+    A visitor class responsible for traversing through each sub-query node, applies
+    entity processors to each node, and returning physical query.
+    """
+
+    def __init__(self, query_settings: QuerySettings) -> None:
+        self.__settings = query_settings
+
+    def _visit_simple_source(self, data_source: Entity) -> ClickhouseQuery:
+        # We are never supposed to get here.
+        raise ValueError("Cannot translate a simple Entity")
+
+    def _visit_join(
+        self, data_source: JoinClause[Entity]
+    ) -> Union[ProcessableQuery[Table], CompositeQuery[Table], JoinClause[Table]]:
```

Comment:
im not really sure what this function does (but I also assume this already existed)

Reply:
this is also probably relevant to my question above about why we need to turn joins into subqueries...